secure-computer-systems/02-design-principles.md

```diff
@@ -0,0 +1,100 @@
+
+#### misc.
+
+This is called economy of mechanism. I am curious the meaning of *mechanism* vs. *feature*, please submit a pull request if you have a good explanation.
+
+Open design is the idea that the attacker should be able to know how the system works and still be unable to execute an attack. The opposite of security by obscurity.
+
+## Least Privilege
+
+Users are authenticated (their identity is verified) and are only allowed to perform certain actions. A user doesn't have to be a person, applications are also considered users in the sense that they have their own privileges/permissions. 
+
+Think about how you have to give applications permissions to do things like look at your pictures on your phone. What would happen if applications had access to all of your photos by default? That sounds like a privacy/security problem.
+
+Different applications have different permissions, and you only give them the permissions they need. This is the principle of least privilege.
+
+### Related Concepts
+
+**Separation of Privileges** is about having more granular ways of controlling access. It makes sense that an application on your phone can separately request access to your camera and to your microphone. This is granularity.
+
+**Fail-safe Defaults** are about denying access by default. The user has no permissions that are not explicitly granted.
+
+
+## Layered Defense or Defense in Depth
+
+Having multiple layers of defense that are diverse will help protect against hackers.
+
+For an example we return to Ken Thompson's Turing Award lecture and discuss a way of protecting ourself from the tricky bug he described.
+
+1. Obtain two independently developed compilers *A* and *B*. Assume that at most one has the bug. Let S<sub>A</sub> and E<sub>A</sub> be the source and executable for *A*. S<sub>B</sub> and E<sub>B</sub> are the source and executable for *B*.
+2. Compile S<sub>A</sub> with E<sub>A</sub> to create *X*. 
+3. Compile S<sub>A</sub> with E<sub>B</sub> to create *Y*.
+4. Compile S<sub>A</sub> with *X* to produce X' as well as with Y to produce Y'.
+5. If not X' == Y' there is a bug in one of the compilers.
+
+The bug Ken Thompson described essentially turns good source code into a bad executable. We have compiled S<sub>A</sub> into two different executables using our compilers and check if the executables produce the same output. If they do not produce the same output then one of the compilers made a bad executable from S<sub>A</sub>.
+
+The defense mechanism described provides an additional layer of security on top of the compiler. You can imagine adding more independent layers for other potential vulnerabilities.
+
+<h2 hid="5ab81562-44e3-49e5-8543-baa4d5f9c237">More Thoughts</h2>
```

Comment:
What is this?

Reply:
I was experimenting with adding interactivity in a way that doesn't impact how it is rendered on GitHub.